### PR TITLE
feat: add Apache Airflow Job MCP tools for Microsoft Fabric

### PR DESCRIPTION
**Pull request by kromerm**
## Summary

Adds full MCP tool support for Microsoft Fabric Apache Airflow Jobs, following the existing 3-layer architecture (Tool, Service, Model).

## New MCP Tools (7)

- list_airflow_jobs - List all Airflow Jobs in a workspace (paginated)
- create_airflow_job - Create a new Airflow Job
- get_airflow_job - Get metadata for a specific Airflow Job
- update_airflow_job - Update display name and/or description
- delete_airflow_job - Delete an Airflow Job (soft or hard delete)
- get_airflow_job_definition - Retrieve the job's DAG/config definition (base64-decoded)
- update_airflow_job_definition - Upload a new definition as JSON (auto base64-encoded)

## Fabric REST API

All operations target: workspaces/{workspaceId}/apacheAirflowJobs

## Files Changed

- Models: 8 new DTOs under Models/AirflowJob/ covering all request/response types including definition parts
- Interface: IFabricAirflowJobService with 7 methods
- Service: FabricAirflowJobService calling Fabric REST API
- Extensions: AirflowJobExtensions.ToFormattedInfo()
- Tool: AirflowJobTool with full validation and error handling
- DI: Registered in ServiceCollectionExtensions

## Tests

- 37 integration tests in AirflowJobToolIntegrationTests.cs
- 31 pass without credentials (validation, auth errors, DI registration)
- 6 SkippableFact authenticated tests including a full lifecycle test (Create, Get, Update, List, GetDefinition, Delete)
- Verified manually against live Fabric API

## Docs

Added docs/airflow-job-feature.md with developer testing guide.

### CODE REVIEW

**Review by Ebram-Tawfik — COMMENTED** [review]
## Code Review

### Fabricated CreatedAt timestamp (Medium)
**File:** AirflowJobTool.cs, line 107

`CreatedAt = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ssZ")`

This is a locally fabricated timestamp, **not** a value returned by the Fabric API. An LLM consuming this tool output will treat it as the authoritative creation time. The Pipeline tool does **not** emit a CreatedAt field - it returns API response properties only.

**Fix:** Remove the CreatedAt property. If the API provides a creation timestamp in the future, add it to the AirflowJob model and map it from the response.

---

### JsonDocument not disposed - resource leak (Low)
**File:** AirflowJobTool.cs, lines 320-321

`JsonDocument.Parse(definitionJson);  // IDisposable, never disposed`

JsonDocument implements IDisposable and rents array buffers internally. Under sustained load this leaks pooled buffers.

**Fix:** Wrap in using:
`using var doc = JsonDocument.Parse(definitionJson);`

---

### Null safety - definition.Parts could NRE (Low)
**File:** AirflowJobTool.cs, lines 275-281

If the API returns `"parts": null`, System.Text.Json will deserialize it as null, overriding the default initializer. This would throw a NullReferenceException caught only by the generic catch, giving an unhelpful error message.

**Fix:**
`PartsCount = definition.Parts?.Count ?? 0`
`Parts = (definition.Parts ?? new()).Select(...)`

---

### Definition API path - please verify (Info)
**File:** FabricAirflowJobService.cs, lines 181-183

Uses `apacheAirflowJobs/{id}/getDefinition` while Pipeline uses `items/{id}/getDefinition`. This is plausible if the Fabric API exposes both patterns, but please verify against the Fabric REST API docs that this is the correct endpoint.

---

### Missing test for null definitionJson (Low)
**File:** AirflowJobToolIntegrationTests.cs

Tests cover empty and invalid JSON, but there is no test for null definitionJson. Consider adding one.

---

### Overall
Well-structured PR that correctly follows the 3-layer architecture. DI registration, validation, error handling, and test coverage are all solid. Recommend addressing the CreatedAt fabrication and JsonDocument disposal before merge.

File: DataFactory.MCP.Core/Tools/AirflowJob/AirflowJobTool.cs [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+            {
+                Success = true,
+                Message = $"Apache Airflow Job '{displayName}' created successfully",
+                AirflowJobId = created.Id,
+                DisplayName = created.DisplayName,
+                Description = created.Description,
+                Type = created.Type,
+                WorkspaceId = created.WorkspaceId,
+                FolderId = created.FolderId,
+                CreatedAt = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ssZ")
```
Comment (Ebram-Tawfik) [inline DataFactory.MCP.Core/Tools/AirflowJob/AirflowJobTool.cs:107]: **Bug (Medium):** This is a locally fabricated timestamp, **not** a value returned by the Fabric API. An LLM consuming this tool output will treat it as the authoritative creation time. The Pipeline tool does not emit a `CreatedAt` field — it returns API response properties only.

**Fix:** Remove the `CreatedAt` property. If the API provides a creation timestamp in the future, add it to the `AirflowJob` model and map it from the response.

File: DataFactory.MCP.Core/Tools/AirflowJob/AirflowJobTool.cs [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+        try
+        {
+            _validationService.ValidateRequiredString(workspaceId, nameof(workspaceId));
+            _validationService.ValidateRequiredString(airflowJobId, nameof(airflowJobId));
+            _validationService.ValidateRequiredString(definitionJson, nameof(definitionJson));
+
+            // Validate JSON format
+            try
+            {
+                JsonDocument.Parse(definitionJson);
```
Comment (Ebram-Tawfik) [inline DataFactory.MCP.Core/Tools/AirflowJob/AirflowJobTool.cs:320]: **Bug (Low):** `JsonDocument` implements `IDisposable` and rents array buffers internally. The parsed document is used only for validation and then abandoned. Under sustained load, this leaks pooled buffers.

**Fix:** Wrap in `using`:
```csharp
using var doc = JsonDocument.Parse(definitionJson);
```

File: DataFactory.MCP.Core/Tools/AirflowJob/AirflowJobTool.cs [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+
+            var definition = await _airflowJobService.GetAirflowJobDefinitionAsync(workspaceId, airflowJobId, format);
+
+            var result = new
+            {
+                Success = true,
+                AirflowJobId = airflowJobId,
+                WorkspaceId = workspaceId,
+                Format = definition.Format,
+                PartsCount = definition.Parts.Count,
```
Comment (Ebram-Tawfik) [inline DataFactory.MCP.Core/Tools/AirflowJob/AirflowJobTool.cs:275]: **Null safety (Low):** If the API returns `"parts": null`, `System.Text.Json` will deserialize it as `null`, overriding the default initializer. This would throw a `NullReferenceException` caught only by the generic catch, giving an unhelpful error message.

**Fix:**
```csharp
PartsCount = definition.Parts?.Count ?? 0,
Parts = (definition.Parts ?? new()).Select(p => new { ... })
```

File: DataFactory.MCP.Core/Services/FabricAirflowJobService.cs [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        string? format = null)
+    {
+        try
+        {
+            ValidateGuids(
+                (workspaceId, nameof(workspaceId)),
+                (airflowJobId, nameof(airflowJobId)));
+
+            var endpoint = FabricUrlBuilder.ForFabricApi()
+                .WithLiteralPath($"workspaces/{workspaceId}/apacheAirflowJobs/{airflowJobId}/getDefinition")
```
Comment (Ebram-Tawfik) [inline DataFactory.MCP.Core/Services/FabricAirflowJobService.cs:182]: **Verify (Info):** This uses `apacheAirflowJobs/{id}/getDefinition` while Pipeline uses `items/{id}/getDefinition`. This is plausible if the Fabric API exposes both patterns, but please verify against the Fabric REST API docs that this is the correct endpoint. If it should be `items/{id}/getDefinition`, every definition call will 404.

File: DataFactory.MCP.Tests/Integration/AirflowJobToolIntegrationTests.cs [hunk 5]
```diff
@@ -0,0 +1,517 @@
+using Xunit;
```
Comment (Ebram-Tawfik) [inline DataFactory.MCP.Tests/Integration/AirflowJobToolIntegrationTests.cs:1]: **Missing test (Low):** Tests cover empty and invalid JSON for `definitionJson`, but there is no test for `null`. Since `definitionJson` is a non-optional `string` parameter, an LLM could pass `null`. Consider adding a test for this case.

**Review by Ebram-Tawfik — COMMENTED** [review]
Overall: Well-structured PR that correctly follows the 3-layer architecture. DI registration, validation, error handling, and test coverage are solid. A few issues to address before merge.

**Review by EbramTawfik — APPROVED** [review]
